Source/WebCore/dom/Observable.h

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2024 Keith Cirkel <webkit@keithcirkel.co.uk>. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#pragma once
+
+#include "AbortSignal.h"
+#include "ScriptWrappable.h"
+#include "SubscribeOptions.h"
+#include "SubscriberCallback.h"
+#include "SubscriptionObserver.h"
+#include "SubscriptionObserverCallback.h"
+#include "VoidCallback.h"
+#include <wtf/RefCounted.h>
+
+namespace WebCore {
+
+class Document;
+class ScriptExecutionContext;
+
+class Observable final : public ScriptWrappable, public RefCounted<Observable> {
+    WTF_MAKE_ISO_ALLOCATED(Observable);
+
+public:
+    using ObserverUnion = std::variant<RefPtr<JSSubscriptionObserverCallback>, SubscriptionObserver>;
```

Comment:
I'd do a `RefPtr<SubscriptionObserverCallback>` rather than the `JS` version unless you've got a good reason for the latter?

Reply:
Refused to compile and I'm not sure exactly how to solve it:

```
In file included from /app/webkit/WebKitBuild/GTK/Debug/JavaScriptCore/PrivateHeaders/JavaScriptCore/ThrowScope.h:28,
/app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.cpp: In lambda function:
/app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.cpp:224:206: error: cannot convert ‘optional<variant<WTF::RefPtr<WebCore::JSSubscriptionObserverCallback, WTF::RawPtrTraits<WebCore::JSSubscriptionObserverCallback>, WTF::DefaultRefDerefTraits<WebCore::JSSubscriptionObserverCallback> >, WebCore::SubscriptionObserver>>’ to ‘optional<variant<WTF::RefPtr<WebCore::SubscriptionObserverCallback, WTF::RawPtrTraits<WebCore::SubscriptionObserverCallback>, WTF::DefaultRefDerefTraits<WebCore::SubscriptionObserverCallback> >, WebCore::SubscriptionObserver>>’
  224 |     RELEASE_AND_RETURN(throwScope, JSValue::encode(toJS<IDLUndefined>(*lexicalGlobalObject, throwScope, [&]() -> decltype(auto) { return impl.subscribe(*context, observerConversionResult.releaseReturnValue(), optionsConversionResult.releaseReturnValue()); })));
/app/webkit/WebKitBuild/GTK/Debug/JavaScriptCore/PrivateHeaders/JavaScriptCore/ExceptionScope.h:126:16: note: in definition of macro ‘RELEASE_AND_RETURN’
  126 |         return expression__; \
In file included from /app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.h:24,
/app/webkit/Source/WebCore/dom/Observable.h:52:45: note:   initializing argument 2 of ‘void WebCore::Observable::subscribe(WebCore::ScriptExecutionContext&, std::optional<std::variant<WTF::RefPtr<WebCore::SubscriptionObserverCallback, WTF::RawPtrTraits<WebCore::SubscriptionObserverCallback>, WTF::DefaultRefDerefTraits<WebCore::SubscriptionObserverCallback> >, WebCore::SubscriptionObserver> >, WebCore::SubscribeOptions)’
   52 |     void subscribe(ScriptExecutionContext&, std::optional<ObserverUnion>, SubscribeOptions);
In file included from /app/webkit/Source/WebCore/bindings/js/JSDOMConvertDictionary.h:29,
/app/webkit/Source/WebCore/bindings/js/JSDOMConvertBase.h: In instantiation of ‘JSC::JSValue WebCore::toJS(JSC::JSGlobalObject&, U&&) [with IDL = IDLUndefined; U = jsObservablePrototypeFunction_subscribeBody(JSC::JSGlobalObject*, JSC::CallFrame*, IDLOperation<JSObservable>::ClassParameter)::<lambda()>]’:
/app/webkit/Source/WebCore/bindings/js/JSDOMConvertBase.h:195:29:   required from ‘JSC::JSValue WebCore::toJS(JSC::JSGlobalObject&, JSC::ThrowScope&, U&&) [with IDL = IDLUndefined; U = jsObservablePrototypeFunction_subscribeBody(JSC::JSGlobalObject*, JSC::CallFrame*, IDLOperation<JSObservable>::ClassParameter)::<lambda()>]’
/app/webkit/WebKitBuild/GTK/Debug/WebCore/DerivedSources/JSObservable.cpp:224:5:   required from here
/app/webkit/Source/WebCore/bindings/js/JSDOMConvertBase.h:156:47: error: incomplete type ‘WebCore::JSConverter<WebCore::IDLUndefined>’ used in nested name specifier
  156 |     return JSConverterOverloader<IDL>::convert(lexicalGlobalObject, std::forward<U>(value));
[15/74] Building CXX object Source/WebCore/CMakeFiles/WebCore.dir/__/__/WebCore/DerivedSources/unified-sources/UnifiedSource-3a52ce78-38.cpp.o
ninja: build stopped: subcommand failed.
```

Reply:
This might be a bug in the code generator and/or support code for the generator or its a limitation of variant. In the generated code for `Observable.subscribe` it seems like it cant cast from `variant<RefPtr<[DerivedClass]>, Other>` to `variant<RefPtr<[BaseClass]>, Other>` which seems wrong. It could be due to the template monstrosity that is created. You should always be able to cast from pointer to derived to pointer to base.

Reply:
So in the mean time this is fine but we should figure out what's going on here.

Reply:
Looks like it might just be a limitation of variant since this doesn't compile:

```C++
#include <variant>

struct A { };
struct B : public A { };

using MyUnion = std::variant<A*, int>;

std::variant<B*, int> makeObject()
{
    return new B();
}

int main()
{
    MyUnion mu = makeObject();
    return 0;
}
```

I'll look into fixing the code generator in the future. I filed this: https://bugs.webkit.org/show_bug.cgi?id=276665

Reply:
I'm hitting this bug while implementing [this](https://drafts.csswg.org/css-view-transitions-2/#extend-document-types) (extension to `document.startViewTransition()`). If nobody is looking into that bugzilla, I could take over it.